### PR TITLE
pkg/asset/manifests/proxy: Create new proxy manifest

### PR DESCRIPTION
**Pull request by patrickdillon**
Add proxy manifest to create config files and Proxy CR when users have a proxy for their cluster.

This PR uses a similar template to other manifests. There is some special logic for handling the NO_PROXY values. First, it generates values that need to be added to the NO_PROXY. Then there is a check to make sure if the user added those values, we will only include it once. I'm not sure if this check is necessary. (There is not currently a check to make sure the user's values are unique, but we could handle that in validation.)

JIRA: CORS-1074

### CODE REVIEW

File: pkg/asset/manifests/proxy.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	"github.com/ghodss/yaml"
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+)
+
+var proxyCfgFilename = filepath.Join(manifestDir, "cluster-proxy-01-config.yml")
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/proxy.go]: nit: `yaml`

File: pkg/asset/manifests/proxy.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+		},
+	}
+
+	return nil
+}
+
+// CreateNoProxy adds platform-specific NO_PROXY values to the user-provided values
+// Platform values are: serviceCIDR, podCIDR, localhost, 127.0.0.1, api.clusterdomain,
+// api-int.clusterdomain, etcd-idx.clusterdomain
+func CreateNoProxy(installConfig *installconfig.InstallConfig) string {
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/proxy.go]: can we make this function private.

File: pkg/asset/manifests/proxy.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	for _, v := range installConfig.Config.Networking.ClusterNetwork {
+		defaults = append(defaults, v.CIDR.String())
+	}
+
+	return CombineValues(installConfig.Config.Proxy.NoProxy, defaults)
+}
+
+// CombineValues filters any duplicates between user- and platform-provided values to create
+// a comma-separated list of unique no-proxy values
+func CombineValues(user string, platform []string) string {
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/proxy.go]: this function is pretty generic and not related to NO_PROXY, can we reword to use no_proxy as example...

also make this function private

File: pkg/asset/manifests/proxy.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+// Platform values are: serviceCIDR, podCIDR, localhost, 127.0.0.1, api.clusterdomain,
+// api-int.clusterdomain, etcd-idx.clusterdomain
+func CreateNoProxy(installConfig *installconfig.InstallConfig) string {
+	defaults := []string{
+		"127.0.0.1",
+		"localhost",
+	}
+
+	defaults = append(defaults, installConfig.Config.Networking.ServiceNetwork[0].String())
+	defaults = append(defaults, fmt.Sprintf("api.%s", installConfig.Config.ClusterDomain()))
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/proxy.go]: use https://github.com/openshift/installer/blob/23aac52885396dec89317e4429b3bc0a1c9398fa/pkg/asset/manifests/utils.go#L36

File: pkg/asset/manifests/proxy.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	defaults = append(defaults, fmt.Sprintf("api-int.%s", installConfig.Config.ClusterDomain()))
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/proxy.go]: use https://github.com/openshift/installer/blob/23aac52885396dec89317e4429b3bc0a1c9398fa/pkg/asset/manifests/utils.go#L40

File: pkg/asset/manifests/proxy.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+// CreateNoProxy adds platform-specific NO_PROXY values to the user-provided values
+// Platform values are: serviceCIDR, podCIDR, localhost, 127.0.0.1, api.clusterdomain,
+// api-int.clusterdomain, etcd-idx.clusterdomain
+func CreateNoProxy(installConfig *installconfig.InstallConfig) string {
+	defaults := []string{
+		"127.0.0.1",
+		"localhost",
+	}
+
+	defaults = append(defaults, installConfig.Config.Networking.ServiceNetwork[0].String())
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/proxy.go]: this should be from the configv1.Network from Networking asset compared to currently the InstallConfig

File: pkg/asset/manifests/proxy.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	defaults = append(defaults, fmt.Sprintf("api.%s", installConfig.Config.ClusterDomain()))
+	defaults = append(defaults, fmt.Sprintf("api-int.%s", installConfig.Config.ClusterDomain()))
+
+	for i := int64(0); i < *installConfig.Config.ControlPlane.Replicas; i++ {
+		etcdHost := fmt.Sprintf("etcd-%d.%s", i, installConfig.Config.ClusterDomain())
+		defaults = append(defaults, etcdHost)
+	}
+
+	for _, v := range installConfig.Config.Networking.ClusterNetwork {
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/proxy.go]: same as https://github.com/openshift/installer/pull/1866/files#r295029977

File: Gopkg.lock [hunk 8]
```diff
@@ -493,18 +493,18 @@
 
 [[projects]]
   branch = "master"
-  digest = "1:38df3f2dd17d211412c30f9fe9656e2a42f3e21d61cf39acf4f536eeef444e01"
+  digest = "1:e0497c97dc8e84143a1f2441358ac201270f6687d8e947b38fc1cea71ab0bee1"
```
Comment (abhinavdahiya) [inline Gopkg.lock]: Can we move all the vendor changes to a separate commit.

**Comment by abhinavdahiya** [comment]
The asset currently is not being pushed to the cluster.. this assets needs to be wired like other config.openshift.io objects like https://github.com/openshift/installer/blob/a99bd59f377226102c146befca60e948f0c601c8/pkg/asset/manifests/operators.go#L124

secondly another commit that updates the assets dep graph https://github.com/openshift/installer/blob/master/docs/design/assetgeneration.md#dependency-graph will be nice.

**Comment by patrickdillon** [comment]
> secondly another commit that updates the assets dep graph https://github.com/openshift/installer/blob/master/docs/design/assetgeneration.md#dependency-graph will be nice.

I have updated the graph. manifests.Proxy shows up but I don't see an installConfig.Proxy. Should installConfig.Proxy be in the graph as well? My branch is up-to-date.

**Comment by patrickdillon** [comment]
```
[padillon@localhost installer]$ cat assets/manifests/cluster-proxy-01-config.yaml 
apiVersion: config.openshift.io/v1
kind: Proxy
metadata:
  creationTimestamp: null
  name: cluster
spec:
  httpProxy: http://padillon:plaintextpassword@example.com
  httpsProxy: https://padillon:plaintextpassword@example.com
  noProxy: cluster.com
status:
  httpProxy: http://padillon:plaintextpassword@example.com
  httpsProxy: https://padillon:plaintextpassword@example.com
  noProxy: localhost,etcd-0.padillon-dr.devcluster.openshift.com,etcd-1.padillon-dr.devcluster.openshift.com,127.0.0.1,https://api.padillon-dr.devcluster.openshift.com:6443,https://api-int.padillon-dr.devcluster.openshift.com:6443,etcd-2.padillon-dr.devcluster.openshift.com,10.128.0.0/14,cluster.com,172.30.0.0/16
```

**Comment by abhinavdahiya** [comment]
> ```
> [padillon@localhost installer]$ cat assets/manifests/cluster-proxy-01-config.yaml 
> apiVersion: config.openshift.io/v1
> kind: Proxy
> metadata:
>   creationTimestamp: null
>   name: cluster
> spec:
>   httpProxy: http://padillon:plaintextpassword@example.com
>   httpsProxy: https://padillon:plaintextpassword@example.com
>   noProxy: cluster.com
> status:
>   httpProxy: http://padillon:plaintextpassword@example.com
>   httpsProxy: https://padillon:plaintextpassword@example.com
>   noProxy: 10.128.0.0/14,127.0.0.1,172.30.0.0/16,cluster.com,etcd-0.padillon-dr.devcluster.openshift.com,etcd-1.padillon-dr.devcluster.openshift.com,etcd-2.padillon-dr.devcluster.openshift.com,https://api-int.padillon-dr.devcluster.openshift.com:6443,https://api.padillon-dr.devcluster.openshift.com:6443,localhost
> ```
> 
> EDIT: added sorting to status.noProxy values

@patrickdillon 
few things pop-out:
`https://api-int.padillon-dr.devcluster.openshift.com:6443` `https://api.padillon-dr.devcluster.openshift.com:6443` are URI, when I think these should have been just the dns names right?

File: pkg/asset/manifests/proxy.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+	return nil
+}
+
+// createNoProxy combines user-provided & platform-specific values to create a comma-separated
+// list of unique NO_PROXY values. Platform values are: serviceCIDR, podCIDR, localhost,
+// 127.0.0.1, api.clusterdomain, api-int.clusterdomain, etcd-idx.clusterdomain
+func createNoProxy(installConfig *installconfig.InstallConfig, network *Networking) string {
+	// use a map to ensure uniqueness
+	noProxySet := map[string]bool{
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/proxy.go]: you can use https://godoc.org/k8s.io/apimachinery/pkg/util/sets which is already imported...

Comment (patrickdillon) [reply pkg/asset/manifests/proxy.go]: This is awesome! thanks. updated.

**Comment by patrickdillon** [comment]
> `https://api-int.padillon-dr.devcluster.openshift.com:6443` `https://api.padillon-dr.devcluster.openshift.com:6443` are URI, when I think these should have been just the dns names right?

I removed the `https` protocol so it now reads: 
```
noProxy: 10.128.0.0/14,127.0.0.1,172.30.0.0/16,api-int.padillon-dr.devcluster.openshift.com:6443,api.padillon-dr.devcluster.openshift.com:6443,cluster.com,etcd-0.padillon-dr.devcluster.openshift.com,etcd-1.padillon-dr.devcluster.openshift.com,etcd-2.padillon-dr.devcluster.openshift.com,localhost
```
I should probably remove the `:6443` port postfix too, right?

Also did you see my question about installConfig.Proxy in the graph: https://github.com/openshift/installer/pull/1866#issuecomment-504046553?

**Comment by abhinavdahiya** [comment]
> > secondly another commit that updates the assets dep graph https://github.com/openshift/installer/blob/master/docs/design/assetgeneration.md#dependency-graph will be nice.
> 
> I have updated the graph. manifests.Proxy shows up but I don't see an installConfig.Proxy. Should installConfig.Proxy be in the graph as well? My branch is up-to-date.

The dependency graph is better assets... Not parts of assets. So I think it's expected that manifest.proxy which is an asset has dependency on entire install config asset.

**Comment by abhinavdahiya** [comment]
> I should probably remove the :6443 port postfix too, right?

Yes.

I think you can net.ParseURL and then use Host field to add to NoProxy list...

File: pkg/asset/manifests/proxy.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	return nil
+}
+
+// createNoProxy combines user-provided & platform-specific values to create a comma-separated
+// list of unique NO_PROXY values. Platform values are: serviceCIDR, podCIDR, localhost,
+// 127.0.0.1, api.clusterdomain, api-int.clusterdomain, etcd-idx.clusterdomain
+func createNoProxy(installConfig *installconfig.InstallConfig, network *Networking) string {
+	set := sets.String{}
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/proxy.go]: nit: if you use sets.NewString() it allows initializing with variable number of string arguments...

**Comment by patrickdillon** [comment]
`noProxy: 10.128.0.0/14,127.0.0.1,172.30.0.0/16,api-int.padillon-dr.devcluster.openshift.com,api.padillon-dr.devcluster.openshift.com,cluster.com,etcd-0.padillon-dr.devcluster.openshift.com,etcd-1.padillon-dr.devcluster.openshift.com,etcd-2.padillon-dr.devcluster.openshift.com,localhost
`
